### ISSUE TITLE
[Feature Request] linux distribution friendly build system

### ISSUE THREAD

**Issue by cdluminate** [pytorch/pytorch#14699]
## Motivation

To decrease the difficulty for linux distribution developers to create package for Pytorch/Caffe2, it would be better if some features could be added to the build system.

## Pitch

My definition of "friendly" is specific to Debian. That means the requirement is the most strict among the distributions... I expect the buildsystem to provide the following flags:

1. `USE_SYSTEM_XXX` flags to use XXX library provided by system, e.g. protobuf instead of building upon the vendored source.
2. `NO_DOWNLOAD` flags to prevent downloading anything during the build (if any).

## Additional context

I'm likely the one who will package Pytorch for Debian/Ubuntu, but I still hesitate to dig into it since the code base is still rapidly changing...

**Comment by ezyang** [comment]
subscribing myself. @ezyang

**Comment by cdluminate** [comment]
We downsteam (Debian) feel it hard to deal with the current build system, when trying to integrate pytorch into Debian official archive. Specifically, Debian does now allow embedded code copies (i.e. those in the `third_party` directory), which means that we have to modify the strange setup.py+CMake build system and let it depend on separately packaged `third_party` libraries.
Currently the build system is rather involved, and we rely on the `USE_SYSTEM_XXX` flags.

Do you have any plan to standarlize the build system of pytorch?

**Comment by ezyang** [comment]
@cdluminate Does it help if I say we'll accept patches that add `USE_SYSTEM_XXX` and `NO_DOWNLOAD` to our cmake. Tag as me as reviewer on them. Also cc'ing @malfet

**Comment by cdluminate** [comment]
> @cdluminate Does it help if I say we'll accept patches that add `USE_SYSTEM_XXX` and `NO_DOWNLOAD` to our cmake. Tag as me as reviewer on them. Also cc'ing @malfet

Thanks! That's a good news. Much better than those deep learning frameworks using Bazel build.

**Comment by cdluminate** [comment]
A question: why do you put two duplicated copies of QNNPACK to the project?
1. `third_pary/QNNPACK` 2. `aten/src/ATen/native/quantized/cpu/qnnpack`

**Comment by ezyang** [comment]
@cdluminate the copy in (2) is a forked version of qnnpack with modified numerics. Since we still have some models using the unforked qnnpack in prod, modifying the upstream copy would require us to fix those prod models, but we're planning on just killing it altogether later.

qnnpack should probably not be packaged for the distro.

**Comment by cdluminate** [comment]
@ezyang Thanks for the pointer. BTW, should we skip packaging any dependency in the following list: NNPACK, XNNPACK, fbgemm, foxi, gloo, ideep ?
Currently my plan only involves packaging the cpu-only version, so we don't need CUDA/ROCm related dependencies currently. (It's not convenient to deal with the proprietary CUDA license from Debian side.)

**Comment by ezyang** [comment]
Yes, this looks reasonable. I do think you should bundle ideep, but I think it's the one that's most likely to warrant its own package, as it's maintained by INtel.

**Comment by cdluminate** [comment]
Hello, I need some more help for understanding onnxifi.

1. The documentation of the [standard onnxifi](https://github.com/onnx/onnx/blob/master/docs/ONNXIFI.md) says "ONNXIFI is a cross-platform API for loading and executing ONNX graphs on optimized backends". Does that mean ONNXIFI is a set of API that can be implemented by different but compatible implementations? Similarly, the BLAS API has many implementations such as openblas, intel MKL, etc. Can I make such an analogy?

2. What's the difference between `foxi` and the "standard" onnxifi from onnx project? Can we smoothly switch the onnxifi backend between `foxi` and the "standard" onnxifi? According to [this response](https://github.com/onnx/onnx/issues/1679#issuecomment-445325260), it sounds like that different backends can work as a drop-in replacement to each other.

3. If my speculatoin is correct, could you please sort the priorities of "foxi", "standard onnxifi", and ["glow onnxifi"](https://github.com/pytorch/glow/blob/master/docs/Onnxifi.md) 

The answers to these questions are important to me. The different mentioned onnxifi implementations (`libonnxifi.so`) will not be able to  coexist under the same `/usr/lib/` directory, and I'll need a priority list to address this coexisting issue.

**Comment by ezyang** [comment]
^- @houseroad

**Comment by joy4eg** [comment]
Yet another thing to [integrade](https://github.com/google/XNNPACK/commit/6cafcdbd6b12e8a405bda9a24db9ed2d2452bf03): `XNNPACK_USE_SYSTEM_LIBS`

**Comment by joy4eg** [comment]
Yet another thing to [integrade](https://github.com/google/XNNPACK/commit/6cafcdbd6b12e8a405bda9a24db9ed2d2452bf03): `XNNPACK_USE_SYSTEM_LIBS`
